### PR TITLE
Several small ManagedHandler fixes

### PR DESCRIPTION
**Pull request by geoffkizer**
(1) Don't check for HTTP 2.0 on response, since we should never see this
(2) Remove IP parsing logic and just let ConnectAsync handle this
(3) Fix proxy auth retry handling
(4) Fix IDNA host name handling in proxy case
(5) Various test additions/fixes

@stephentoub @Priya91 @wfurt @davidsh

### CODE REVIEW

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+using System.Collections.Generic;
+using System.Net.Test.Common;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Http.Functional.Tests
+{
+    public abstract class IdnaProtocolTests : HttpClientTestBase
+    {
+        private const string s_IdnaHost = "âbçdë.com";
```
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: We try not to add Unicode characters to source files.  Instead, we use "\uXXXX" syntax in the string literals.

Comment (Priya91) [inline src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: We scrubbed our source code during opensource to escape unicode chars in our files.. We should maintain consistency..

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                _proxyUri = proxyUri;
+            }
+
+            public ICredentials Credentials { get => null; set => throw new NotImplementedException(); }
+            public Uri GetProxy(Uri destination) => _proxyUri;
+            public bool IsBypassed(Uri host) => false;
+        }
+    }
+
+    public sealed class DefaultHandler_IdnaProtocolTests : IdnaProtocolTests
```
Comment (Priya91) [inline src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: The other test files currently use naming as HttpClientHandler_

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    {
+    }
+
+    public sealed class ManagedHandler_IdnaProtocolTests : IdnaProtocolTests
```
Comment (Priya91) [inline src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: This should go in the ManagedHandlerTests.cs file, all our managed handler tests are organized there.

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs (hunk 1 above)
Comment (Priya91) [reply src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: Should we have more unicode test cases?

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs [hunk 4]
```diff
@@ -226,9 +227,16 @@ private Task WriteFormattedInt32Async(int value, CancellationToken cancellationT
...
-                await WriteStringAsync(
-                    _usingProxy ? request.RequestUri.AbsoluteUri : request.RequestUri.PathAndQuery,
-                    cancellationToken).ConfigureAwait(false);
+
+                if (_usingProxy)
+                {
+                    // Proxied requests contain full URL
+                    Debug.Assert(request.RequestUri.Scheme == Uri.UriSchemeHttp);
+                    await WriteBytesAsync(s_httpSchemeAndDelimiter, cancellationToken).ConfigureAwait(false);
+                    await WriteAsciiStringAsync(request.RequestUri.IdnHost, cancellationToken).ConfigureAwait(false);
```
Comment (Priya91) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: Should we use the Host header if present?

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Http.Functional.Tests
+{
+    public abstract class IdnaProtocolTests : HttpClientTestBase
+    {
+        private const string s_IdnaHost = "âbçdë.com";
+        private static Uri s_IdnaUri = new Uri($"http://{s_IdnaHost}/");
+        private static string s_EncodedHostName = s_IdnaUri.IdnHost;
```
Comment (Priya91) [inline src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: Also all names s_camelCasing

File: src/System.Net.Http/src/System/Net/Http/Managed/ConnectHelper.cs [hunk 6]
```diff
@@ -22,10 +22,7 @@ internal static class ConnectHelper
                 // to cancel it if needed.
                 using (var saea = new BuilderAndCancellationTokenSocketAsyncEventArgs(cancellationToken))
                 {
-                    // Configure which server to which to connect.
-                    saea.RemoteEndPoint = IPAddress.TryParse(host, out IPAddress address) ?
-                        (EndPoint)new IPEndPoint(address, port) :
-                        new DnsEndPoint(host, port);
+                    saea.RemoteEndPoint = new DnsEndPoint(host, port);
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Managed/ConnectHelper.cs]: While the change works, if an address is specified, doesn't that end up taking a non-trivially more expensive code path? Before this is changed, it'd be good to look at a profile (including allocations); if I'm right, maybe we should still take the change and fix it in Socket, but I wouldn't want to regress this when the original code was pretty simple.

File: src/Common/tests/System/Net/Http/LoopbackServer.cs [hunk 7]
```diff
@@ -153,6 +153,7 @@ private static string ComputeWebSocketHandshakeSecurityAcceptValue(string secWeb
         {
             options = options ?? new Options();
             Socket s = await server.AcceptAsync().ConfigureAwait(false);
+            s.NoDelay = true;
```
Comment (stephentoub) [inline src/Common/tests/System/Net/Http/LoopbackServer.cs]: Did this affect something tangibly?

File: src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs [hunk 8]
```diff
@@ -123,7 +123,7 @@ private sealed class DribbleStream : Stream
                 for (int i = 0; i < count; i++)
                 {
                     await _wrapped.WriteAsync(buffer, offset + i, 1);
-                    await Task.Yield(); // introduce short delays, enough to send packets individually but so long as to extend test duration significantly
+                    await Task.Delay(3); // introduce short delays, enough to send packets individually but not so long as to extend test duration significantly
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs]: This is going to add appx 10-15ms per byte, at least on Windows. What impact does that have on test running time? Why was this change needed? And where did "3" come from? I'd expect no difference on Windows btw 1 and 3, add it's less than a default timer quantum.

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+                    HttpResponseMessage response = await getResponseTask;
+                    Console.WriteLine($"Location header={response.Headers.Location}");
+                    Console.WriteLine($"Location header.Host={response.Headers.Location.Host}");
+                    Console.WriteLine($"Location header.IdnHost={response.Headers.Location.IdnHost}");
+                }
+            });
+        }
+
+        sealed class SimpleWebProxy : IWebProxy
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: Don't we have one of these already somewhere? And could we just use WebProxy?

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Net.Test.Common;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Http.Functional.Tests
+{
+    public abstract class IdnaProtocolTests : HttpClientTestBase
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: As abstract, this won't test WinHttpHandler or CurlHandler; is that intentional?

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs (hunk 9 above)
Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: Probably, yeah.  I'll fix.

File: src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs (hunk 8 above)
Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs]: It didn't seem to have a measurable effect on running time.  The dribble stream stuff is only used for a few tests.

I made the change speculatively; I can back out if you think it's not useful.  I honestly don't know if it matters at all.  If it does, Yield is probably sufficient here.

File: src/System.Net.Http/src/System/Net/Http/Managed/ConnectHelper.cs (hunk 6 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/Managed/ConnectHelper.cs]: Yeah, you are correct here.  I should have looked at the ConnectAsync code more closely.

We should consider fixing this in Socket, but that seems like a more involved change.

I'll revert and file a separate issue.

File: src/Common/tests/System/Net/Http/LoopbackServer.cs (hunk 7 above)
Comment (geoffkizer) [reply src/Common/tests/System/Net/Http/LoopbackServer.cs]: No, but it seems pretty broken to not set NoDelay.

File: src/System.Net.Http/src/System/Net/Http/Managed/ConnectHelper.cs (hunk 6 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/Managed/ConnectHelper.cs]: Filed #26517

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs (hunk 2 above)
Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: Thanks, will fix

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs (hunk 4 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: I don't think so.  It's possible for the user to specify an IP address in the Uri and also specify an explicit Host header; the IP address is used to connect.  Presumably we want the same behavior for proxies, so we should send the host from the RequestUri.

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs (hunk 10 above)
Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: There's a derived handler at the bottom of the file for WinHttpHandler/CurlHandler.  As Priya pointed out, this should probably be called HttpClientHandler_IdnaProtocolTests; its currently called DefaultHandler_IdnaProtocolTests, but I'll fix.

Also, as Priya pointed out above, this pattern is inconsistent with current practice; we currently have all the derived ManagedHandler_*Tests classes in ManagedHandlerTest.cs, and the base classes are not abstract.  I don't love this pattern.  I'd like to see us do something like some of the sockets tests do, where the pattern is an abstract base and derived classes for each variation.  This makes it easier to customize behavior per-variation, which I suspect we will want more of in the future, beyond just UseManagedHandler = true.

That said, I'm happy to revert to the existing pattern for now if that's what folks think is best.

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs (hunk 3 above)
Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: See my response to Stephen below

File: src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs (hunk 8 above)
Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs]: > It didn't seem to have a measurable effect on running time.

I'm surprised, but if it really doesn't have a measurable impact, it's fine.

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs (hunk 1 above)
Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: Is Unicode allowed in comments?  I'm guessing not.

Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: Probably.  I'll add a few more.

**Comment by geoffkizer** [comment]
I think I addressed all the issues except the test class structure one.  Let me know if I missed anything.

File: src/Common/tests/System/Net/Http/LoopbackServer.cs (hunk 7 above)
Comment (stephentoub) [reply src/Common/tests/System/Net/Http/LoopbackServer.cs]: I'm fine with seeing it, but why is that "broken"? Just curious.

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs [hunk 11]
```diff
@@ -519,7 +527,6 @@ private void ParseStatusLine(Span<byte> line, HttpResponseMessage response)
             response.Version =
                 (majorVersion == '1' && minorVersion == '1') ? HttpVersionInternal.Version11 :
                 (majorVersion == '1' && minorVersion == '0') ? HttpVersionInternal.Version10 :
-                (majorVersion == '2' && minorVersion == '0') ? HttpVersionInternal.Version20 :
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: No strong opinion, but should the response.Version accurately reflect what was sent back by the server, even if it was erroneous? I'm wondering if in addition to removing this line, we should change the final else to just construct an HttpVersion with the supplied values.

Also, I think we were using HttpVersion call because of 2.0. Now that it's being removed, can we switch to using HttpVersion instead?

**Comment by geoffkizer** [comment]
@dotnet/dnceng Why aren't the test failures for netfx being logged and reported correctly?

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        protected abstract bool SupportsIdna { get; }
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: Nit: you could make it virtual with => true and then not have to override it for true later. That's generally what's done elsewhere.

**Comment by MattGal** [comment]
@geoffkizer looking at this now.

**Comment by geoffkizer** [comment]
Re test file structure:  It looks like we rely on the managed tests being in ManagedHandlerTest.cs so they don't get compiled on netfx.

I think I'll just use the existing structure here.  Still don't love this but it works.

File: src/Common/tests/System/Net/Http/LoopbackServer.cs (hunk 7 above)
Comment (geoffkizer) [reply src/Common/tests/System/Net/Http/LoopbackServer.cs]: I should clarify: Without NoDelay, the *dribble tests* are pretty broken.  The kernel will buffer sends so we won't actually send the bytes one at a time, which is the point of the dribble test stuff.

The loopback server is not broken in general, though it does seem slightly better to use NoDelay, just because kernel buffering may produce different behavior than we're expecting.

**Comment by geoffkizer** [comment]
@MattGal Thanks.  I did find the test results in "view as plain text", though not in "console output" which is where I originally looked.

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs (hunk 11 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: Yeah, I agree.  That said, it's not a huge issue, as in practice it should never actually happen... but still seems like a good thing to do.

> Also, I think we were using HttpVersion call because of 2.0. Now that it's being removed, can we switch to using HttpVersion instead?

Do you mean, we should switch from HttpVersionInternal to HttpVersion?  Yeah, we can do that.  I filed an issue on this somewhere...

File: src/Common/tests/System/Net/Http/LoopbackServer.cs (hunk 7 above)
Comment (stephentoub) [reply src/Common/tests/System/Net/Http/LoopbackServer.cs]: Ok

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs (hunk 11 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: > Do you mean, we should switch from HttpVersionInternal to HttpVersion

Sorry, yes, writing with non dominant hand on phone :)

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs (hunk 1 above)
Comment (weitzhandler) [reply src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: Maybe replace with `\u` prefixes.

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs (hunk 11 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: In reading the RFC (https://tools.ietf.org/html/rfc7230#section-2.6), I think we should be a bit more restrictive here.  In particular, we should fail if major = 0 (which is undefined) or major >= 2 (which implies incompatibility with 1.x).  

So that means that, for example, 1.2 is still valid and we should construct a valid response Version for this.

Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: It looks like @caesar1995 is making some related test changes here.

**Comment by geoffkizer** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please
@dotnet-bot test Outerloop Linux x64 Debug Build please
@dotnet-bot test Outerloop OSX x64 Debug Build please

File: src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs (hunk 1 above)
Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/IdnaProtocolTests.cs]: The updated commit uses \u.

**Comment by mmitche** [comment]
@geoffkizer The default console output view only shows a limited number of lines.  If you go to the top (or side) there is a full view that isn't the text view.  The text and console views come from the same log file.

File: src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs (hunk 8 above)
Comment (pjanotti) [reply src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs]: Since you mention dribble tests... I had two failures with them but they looked like the thing that `LoopbackServer` closed the connection socket under the client (`System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host`), one of the issues on #25880.
